docs/error-handling.md

```diff
@@ -0,0 +1,190 @@
+inputs from client can cause ERRORs. For example, if a client tries to
+call a mgmt API that doesn't exist, or if a compute node sends passes
+an LSN that has already been garbage collected away.
+
+These should *not* happen during normal operations. "Normal
+operations" is not a very precise concept. But for example, disk
+errors is not expected to happen when the system is working, so that
+counts as an Error. However, if a TCP connection to a compute node is
+lost, that is not considered an Error, because it doesn't affect the
+pageserver's or safekeeper's operation in any way, and happens fairly
+frequently when compute nodes are shut down, or are killed abruptly
+because of errors in the compute.
+
+**Errors are monitored, and always need human investigation to determine
+the cause.**
+
+Whether something should be logged at ERROR, WARNING or INFO level can
+depend on the callers and clients. For example, it might be unexpected
+and a sign of a serious issue, if the console calls the
+"timeline_detail" mgmt API for a timeline that doesn't exist. ERROR
+would be appropriate in that case. But if the console routinely calls
+the API after deleting a timeline, to check if the deletion has
+completed, then it would be totally normal and INFO or DEBUG level
+would be more appropriate. If a message is logged as an ERROR, but it
+in fact happens frequently in production and never requires any
+action, it should probably be demoted to INFO level.
+
+### Warn
+
+Examples:
+- could not remove temporary file "foobar.temp"
+- unrecognized file "foobar" in timeline directory
+
+Warnings are similar to Errors, in that they should not happen
+when the system is operating normally. The difference between Error and
+Warning is that an Error means that the operation failed, whereas Warning
+means that something unexpected happened, but the operation continued anyway.
+For example, if deleting a file fails because the file already didn't exist,
+it should be logged as Warning.
+
+> **Note:** The python regression tests, under `test_regress`, check the
+> pageserver log after each test for any ERROR and WARN lines. If there are
+> any ERRORs or WARNs that have not been explicitly listed in the test as
+> allowed, the test is marked a failed. This is to catch unexpected errors
+> e.g. in background operations, that don't cause immediate misbehaviour in
+> the tested functionality.
+
+### Info
+
+Info level is used to log useful information when the system is
+operating normally. Info level is appropriate e.g. for logging state
+changes, background operations, and network connections.
+
+Examples:
+- "system is shutting down"
+- "tenant was created"
+- "retrying S3 upload"
+
+### Debug & Trace
+
+Debug and Trace level messages are not Debug-level messages are not
+printed to the log in our normal production configuration, but could
+be enabled for a specific server or tenant, to aid debugging.
+(Although we don't actually have that capability as of this writing).
```

Comment:
It'd be good to have more information on what the distinction between Debug and Trace should be, and when to use them instead of Info

Reply:
There was a brief discussion whether we should adopt https://github.com/trentm/node-bunyan#levels since there would have been a rust library following the same scheme. Although didn't necessarily warm up to idea to introduce one more level, because that would mean going through pretty much any existing error to figure out if it's error or fatal, the bunyan definitions could have some useful points.

Reply:
I like those node-bunyan definitions, and an additional FATAL level also makes sense to me. However, `tracing` doesn't have FATAL, so I don't think we can adopt that.